File: sdks/go/snapshots.go
package fuse

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"net/http"
	"net/url"
)

type SnapshotsService struct {
	t *transport
}

func newSnapshotsService(t *transport) *SnapshotsService {
	return &SnapshotsService{t: t}
}

type ListSnapshotsOptions struct {
	VMID     string
	TaskID   string
	TenantID string
	State    string

	// Mode narrows to one creation mode, e.g. "build" for build artifacts.
	Mode string

	// Name matches the snapshot's metadata name, the lookup key `fuse build`
	// sets on an artifact.
	Name string

	// Limit caps the page size (server default 50, max 200; <=0 uses the
	// server default). Only consulted by ListPage — List always requests
	// the server's max page size internally since it walks every page.
	Limit int

	// Cursor resumes after a previous page's SnapshotPage.NextCursor. It is
	// an opaque token; treat it as such rather than parsing it.
	Cursor string
}

// SnapshotPage is one page of a List call, plus the cursor to fetch the
// next one.
type SnapshotPage struct {
	Snapshots []Snapshot `json:"snapshots"`
	// NextCursor is empty once there are no more results.
	NextCursor string `json:"next_cursor,omitempty"`
}

func (s *SnapshotsService) Create(ctx context.Context, vmID string, reqBody SnapshotRequest) (*Snapshot, error) {
	if s == nil || s.t == nil {
		return nil, errors.New("snapshots service is not configured")
	}
	if vmID == "" {
		return nil, errors.New("vm id is required")
	}
	path := "/v1/environments/" + url.PathEscape(vmID) + "/snapshots"
	req, err := s.t.newRequest(ctx, http.MethodPost, path, nil, reqBody)
	if err != nil {
		return nil, err
	}
	resp, err := s.t.do(req)
	if err != nil {
		return nil, err
	}
	if err := CheckResponse(resp); err != nil {
		return nil, err
	}
	defer func() { _ = resp.Body.Close() }()
	var snap Snapshot
	if err := json.NewDecoder(resp.Body).Decode(&snap); err != nil {
		return nil, fmt.Errorf("decode snapshot: %w", err)
	}
	return &snap, nil
}

// List returns every snapshot matching opt, transparently walking every
// result page. For explicit single-page control (e.g. a CLI --cursor
// flag), use ListPage.
func (s *SnapshotsService) List(ctx context.Context, opt ListSnapshotsOptions) ([]Snapshot, error) {
	var out []Snapshot
	cursor := opt.Cursor
	for {
		page, err := s.ListPage(ctx, ListSnapshotsOptions{
			VMID:     opt.VMID,
			TaskID:   opt.TaskID,
			TenantID: opt.TenantID,
			State:    opt.State,
			Mode:     opt.Mode,
			Name:     opt.Name,
			Limit:    maxPageLimit,
			Cursor:   cursor,
		})
		if err != nil {
			return nil, err
		}
		out = append(out, page.Snapshots...)
		if page.NextCursor == "" {
			break
		}
		cursor = page.NextCursor
	}
	return out, nil
}

// ListPage returns one page of snapshots matching opt.
func (s *SnapshotsService) ListPage(ctx context.Context, opt ListSnapshotsOptions) (*SnapshotPage, error) {
	if s == nil || s.t == nil {
		return nil, errors.New("snapshots service is not configured")
	}
	values := url.Values{}
	if opt.VMID != "" {
		values.Set("vm_id", opt.VMID)
	}
	if opt.TaskID != "" {
		values.Set("task_id", opt.TaskID)
	}
	if opt.TenantID != "" {
		values.Set("tenant_id", opt.TenantID)
	}
	if opt.State != "" {
		values.Set("state", opt.State)
	}
	if opt.Mode != "" {
		values.Set("mode", opt.Mode)
	}
	if opt.Name != "" {
		values.Set("name", opt.Name)
	}
	setPaginationParams(values, opt.Limit, opt.Cursor)
	req, err := s.t.newRequest(ctx, http.MethodGet, "/v1/snapshots", values, nil)
	if err != nil {
		return nil, err
	}
	resp, err := s.t.do(req)
	if err != nil {
		return nil, err
	}
	if err := CheckResponse(resp); err != nil {
		return nil, err
	}
	defer func() { _ = resp.Body.Close() }()
	var out snapshotList
	if err := json.NewDecoder(resp.Body).Decode(&out); err != nil {
		return nil, fmt.Errorf("decode snapshots: %w", err)
	}
	page := &SnapshotPage{Snapshots: out.Snapshots}
	if out.NextCursor != nil {
		page.NextCursor = *out.NextCursor
	}
	return page, nil
}

func (s *SnapshotsService) Get(ctx context.Context, snapshotID string) (*Snapshot, error) {
	if s == nil || s.t == nil {
		return nil, errors.New("snapshots service is not configured")
	}
	if snapshotID == "" {
		return nil, errors.New("snapshot id is required")
	}
	path := "/v1/snapshots/" + url.PathEscape(snapshotID)
	req, err := s.t.newRequest(ctx, http.MethodGet, path, nil, nil)
	if err != nil {
		return nil, err
	}
	resp, err := s.t.do(req)
	if err != nil {
		return nil, err
	}
	if err := CheckResponse(resp); err != nil {
		return nil, err
	}
	defer func() { _ = resp.Body.Close() }()
	var snap Snapshot
	if err := json.NewDecoder(resp.Body).Decode(&snap); err != nil {
		return nil, fmt.Errorf("decode snapshot: %w", err)
	}
	return &snap, nil
}

func (s *SnapshotsService) Delete(ctx context.Context, snapshotID string) error {
	if s == nil || s.t == nil {
		return errors.New("snapshots service is not configured")
	}
	if snapshotID == "" {
		return errors.New("snapshot id is required")
	}
	path := "/v1/snapshots/" + url.PathEscape(snapshotID)
	req, err := s.t.newRequest(ctx, http.MethodDelete, path, nil, nil)
	if err != nil {
		return err
	}
	resp, err := s.t.do(req)
	if err != nil {
		return err
	}
	if err := CheckResponse(resp); err != nil {
		return err
	}
	return nil
}

func (s *SnapshotsService) Restore(ctx context.Context, snapshotID string) error {
	if s == nil || s.t == nil {
		return errors.New("snapshots service is not configured")
	}
	if snapshotID == "" {
		return errors.New("snapshot id is required")
	}
	path := "/v1/snapshots/" + url.PathEscape(snapshotID)
	values := url.Values{}
	values.Set("action", "restore")
	req, err := s.t.newRequest(ctx, http.MethodPost, path, values, nil)
	if err != nil {
		return err
	}
	resp, err := s.t.do(req)
	if err != nil {
		return err
	}
	if err := CheckResponse(resp); err != nil {
		return err
	}
	return nil
}

File: sdks/python/src/fuse/_services/snapshots.py
from __future__ import annotations

from typing import Optional
from urllib.parse import quote

from .._transport import Transport
from ..types import Snapshot, SnapshotPage, SnapshotRequest

# the server's max page size; list() requests this internally so it walks
# every page in as few round trips as possible.
_MAX_PAGE_LIMIT = 200


class SnapshotsService:
    def __init__(self, transport: Transport) -> None:
        self._t = transport

    def create(
        self, vm_id: str, request: Optional[SnapshotRequest] = None
    ) -> Snapshot:
        if not vm_id:
            raise ValueError("vm id is required")
        path = f"/v1/environments/{quote(vm_id, safe='')}/snapshots"
        resp = self._t.request("POST", path, body=request or SnapshotRequest())
        return Snapshot.model_validate(resp.json())

    def list(
        self,
        *,
        vm_id: str = "",
        task_id: str = "",
        tenant_id: str = "",
        state: str = "",
        layer_key: str = "",
        arch: str = "",
        cursor: str = "",
    ) -> list[Snapshot]:
        # returns every snapshot matching the filters, transparently walking
        # every result page. for explicit single-page control (e.g. a cursor
        # from a previous call), use list_page.
        out: list[Snapshot] = []
        while True:
            page = self.list_page(
                vm_id=vm_id,
                task_id=task_id,
                tenant_id=tenant_id,
                state=state,
                layer_key=layer_key,
                arch=arch,
                limit=_MAX_PAGE_LIMIT,
                cursor=cursor,
            )
            out.extend(page.snapshots)
            if not page.next_cursor:
                break
            cursor = page.next_cursor
        return out

    def list_page(
        self,
        *,
        vm_id: str = "",
        task_id: str = "",
        tenant_id: str = "",
        state: str = "",
        layer_key: str = "",
        arch: str = "",
        limit: int = 0,
        cursor: str = "",
    ) -> SnapshotPage:
        # returns one page of snapshots matching the filters.
        #
        # layer_key narrows to the build layers taken after one setup step,
        # and arch to the artifacts built on one architecture ("amd64",
        # "arm64"). arch is a separate filter rather than part of the layer
        # key because a rootfs is not portable across architectures, so a
        # layer lookup that does not constrain arch can be served bytes it
        # cannot boot. an empty filter is dropped by clean_params, so it means
        # "do not filter" rather than "match artifacts with no layer key".
        params: dict[str, str] = {
            "vm_id": vm_id,
            "task_id": task_id,
            "tenant_id": tenant_id,
            "state": state,
            "layer_key": layer_key,
            "arch": arch,
        }
        if limit > 0:
            params["limit"] = str(limit)
        if cursor:
            params["cursor"] = cursor
        resp = self._t.request("GET", "/v1/snapshots", params=params)
        return SnapshotPage.model_validate(resp.json())

    def get(self, snapshot_id: str) -> Snapshot:
        if not snapshot_id:
            raise ValueError("snapshot id is required")
        resp = self._t.request("GET", f"/v1/snapshots/{quote(snapshot_id, safe='')}")
        return Snapshot.model_validate(resp.json())

    def delete(self, snapshot_id: str) -> None:
        if not snapshot_id:
            raise ValueError("snapshot id is required")
        self._t.request("DELETE", f"/v1/snapshots/{quote(snapshot_id, safe='')}")

    def restore(self, snapshot_id: str) -> None:
        if not snapshot_id:
            raise ValueError("snapshot id is required")
        path = f"/v1/snapshots/{quote(snapshot_id, safe='')}"
        self._t.request("POST", path, params={"action": "restore"})
